primitives/consensus/common/src/import_queue/basic_queue.rs

```diff
@@ -240,9 +248,30 @@ impl<B: BlockT, Transaction: Send> BlockImportWorker<B, Transaction> {
 		blocks: Vec<IncomingBlock<B>>
 	) -> impl Future<Output = (BoxBlockImport<B, Transaction>, V)> {
 		let mut result_sender = self.result_sender.clone();
+		let metrics = self.metrics.clone();
 
 		import_many_blocks(block_import, origin, blocks, verifier, self.delay_between_blocks)
 			.then(move |(imported, count, results, block_import, verifier)| {
+				if let Some(metrics) = metrics {
+					let updates = results.iter().fold(HashMap::new(), |mut acc, result| {
```

Comment:
What do you think of doing it this way:
```
let mut counter = [0u64; 8];

match result.0 {
		Err(BlockImportError::IncompleteHeader(_)) => counter[0] += 1,
		Err(BlockImportError::VerificationFailed(_,_)) => counter[1] += 1,
		Err(BlockImportError::BadBlock(_)) => counter[2] += 1,
		Err(BlockImportError::MissingState) => counter[3] += 1,
		Err(BlockImportError::UnknownParent) => counter[4] += 1,
		...
}
```

If we do this, we don't need any allocations.